File: src/com/kk_electronic/kkportal/core/security/Digest.java
/*
 * Copyright 2010 kk-electronic a/s. 
 * 
 * This file is part of KKPortal.
 *
 * KKPortal is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * KKPortal is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with KKPortal.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
package com.kk_electronic.kkportal.core.security;

import com.google.gwt.core.client.GWT;
import com.google.gwt.user.client.rpc.AsyncCallback;
import com.google.inject.Inject;
import com.kk_electronic.kkportal.core.rpc.Request;

/**
 * 
 * @author Jes Andersen
 */
public class Digest implements SecurityMethod {

	private final Hasher hasher;
	private final String username;
	private final String secret;

	@Inject
	public Digest(Hasher hasher) {
		this.hasher = hasher;
		this.username = "jes";
		this.secret = hasher.hash(username + ':' + "1234");
	}
	
	@Override
	public void sign(Request request, AsyncCallback<Request> asyncCallback) {
		String requestSignature = request.getSignature();
		GWT.log("DIGEST-Signing: " + requestSignature);
		String digestinput = secret+':'+requestSignature;
		request.setParams(new Object[]{
				hasher.hash(digestinput),
				username,
				request.getParams()
		});
		asyncCallback.onSuccess(request);
	}
}

File: server/w2b/security/Digest.py
#
# Copyright 2010 kk-electronic a/s. 
# 
# This file is part of KKPortal.
#
# KKPortal is free software: you can redistribute it and/or modify it
# under the terms of the GNU Lesser General Public License as published
# by the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# KKPortal is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with KKPortal.  If not, see <http://www.gnu.org/licenses/>.
#
'''
Created on Aug 3, 2010

@author: albatros
'''
import hashlib
import simplejson as json
from w2b.database.portal import users

_db = {
       "jes":hashlib.sha256('jes:bb.windpowerhub.com:secret').hexdigest()
       }

def getUser(username):
    return users.select(users.c.name == username).execute().fetchone() #@UndefinedVariable

def getRequestSignature(method,params,id):
    return ",".join((str(id),method,json.dumps(params,separators=(',',':'),sort_keys=True)))

def process(method,params,id):
    if isinstance(params, list) and len(params) > 1:
        providedDigest,username,realparams = params
        requestSignature = getRequestSignature(method, realparams, id)
        user = getUser(username)
        if user is None:
            return False
        realDigest = hashlib.sha256(user.secret + ':' + requestSignature).hexdigest()
        if isinstance(providedDigest, str) and providedDigest == realDigest:
            params[:] = realparams
            return True
    return False